src/wp-includes/block-template-utils.php

```diff
@@ -957,34 +1003,10 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
 	}
 
 	if ( ! isset( $query['wp_id'] ) ) {
-		$template_files = _get_block_templates_files( $template_type );
+		$user_templates = wp_list_pluck( $query_result, 'id' );
+		$template_files = _get_block_templates_files( $template_type, $query, $user_templates );
 		foreach ( $template_files as $template_file ) {
-			$template = _build_block_template_result_from_file( $template_file, $template_type );
-
-			if ( $post_type && ! $template->is_custom ) {
-				continue;
-			}
-
-			if ( $post_type &&
-				isset( $template->post_types ) &&
-				! in_array( $post_type, $template->post_types, true )
-			) {
-				continue;
-			}
-
-			$is_not_custom   = false === array_search(
-				get_stylesheet() . '//' . $template_file['slug'],
-				wp_list_pluck( $query_result, 'id' ),
-				true
-			);
-			$fits_slug_query =
-				! isset( $query['slug__in'] ) || in_array( $template_file['slug'], $query['slug__in'], true );
-			$fits_area_query =
-				! isset( $query['area'] ) || $template_file['area'] === $query['area'];
-			$should_include  = $is_not_custom && $fits_slug_query && $fits_area_query;
-			if ( $should_include ) {
-				$query_result[] = $template;
-			}
+			$query_result[] = _build_block_template_result_from_file( $template_file, $template_type );
```

Comment:
This is the key change of this PR: instead of building the template object _for all items_ and filter them _after_, we filter _before_ this step, so we have to build less items. We can do this because all the data used to filter the items is already available at `_get_block_templates_files`.